sw/device/tests/sim_dv/keymgr_key_derivation.c

```diff
@@ -0,0 +1,278 @@
+const uint32_t kMaxVerBl0 = 2;
+
+const test_config_t kTestConfig;
+
+/**
+ * Writes `size` words of `data` into flash info page.
+ *
+ * @param page_id Info page ID to write to.
+ * @param data Data to write.
+ * @param size Number of 4B words to write from `data` buffer.
+ */
+static void write_info_page(uint32_t page_id, const uint32_t *data,
+                            size_t size) {
+  mp_region_t info_region = {.num = page_id,
+                             .base = 0x0,  // only used to calculate bank id.
+                             .size = 0x1,  // unused for info pages.
+                             .part = kInfoPartition,
+                             .rd_en = true,
+                             .prog_en = true,
+                             .erase_en = true,
+                             .scramble_en = false};
+  flash_cfg_region(&info_region);
+
+  uint32_t address = FLASH_MEM_BASE_ADDR + page_id * flash_get_page_size();
+
+  ASSERT_EQZ(flash_page_erase(address, kInfoPartition));
+  ASSERT_EQZ(flash_write(address, kInfoPartition, data, size));
+
+  for (size_t i = 0; i < size; ++i) {
+    uint32_t got_data;
+    ASSERT_EQZ(flash_read(address + i * sizeof(uint32_t), kInfoPartition,
+                          /*size=*/1, &got_data));
+    CHECK(got_data == data[i]);
+  }
+  info_region.rd_en = false;
+  info_region.prog_en = false;
+  info_region.erase_en = false;
+  flash_cfg_region(&info_region);
+}
+
+static void init_flash(void) {
+  // setup default access for data partition
+  flash_default_region_access(/*rd_en=*/true, /*prog_en=*/true,
+                              /*erase_en=*/true);
+  // Initialize flash secrets.
+  write_info_page(kFlashInfoPageIdCreatorSecret, kCreatorSecret,
+                  ARRAYSIZE(kCreatorSecret));
+  write_info_page(kFlashInfoPageIdOwnerSecret, kOwnerSecret,
+                  ARRAYSIZE(kOwnerSecret));
+}
+
+/**
+ * Busy-wait until the DAI is done with whatever operation it is doing.
+ */
+static dif_otp_ctrl_t otp;
+
+/**
+ * Lock otp secret 2 partition so that on reboot flash seeds can be
+ * automatically loaded
+ */
+static void lock_otp_secret_partition2(void) {
+  // initialize otp
+  CHECK_DIF_OK(dif_otp_ctrl_init(
+      mmio_region_from_addr(TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR), &otp));
+
+  otp_ctrl_testutils_lock_partition(&otp, kDifOtpCtrlPartitionSecret2, 0);
+}
+
+/** Place kmac into sideload mode for correct keymgr operation */
+static void init_kmac_for_keymgr(void) {
+  dif_kmac_t kmac;
+  CHECK_DIF_OK(
+      dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR), &kmac));
+
+  // Configure KMAC hardware using software entropy.
+  dif_kmac_config_t config = (dif_kmac_config_t){
+      .sideload = true,
+  };
+  CHECK_DIF_OK(dif_kmac_configure(&kmac, config));
+
+  for (uint8_t i = 0; i < KMAC_PREFIX_SIZE; ++i) {
+    mmio_region_write32(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR),
+                        KMAC_PREFIX_0_REG_OFFSET + i * 4, kKmacPrefix[i]);
+  }
```

Comment:
Is this needed in keymgr mode?

Reply:
yes, this is needed. Based on kmac spec, we should at least configure the `prefix` with these value for the first a few bytes.

> The [NIST SP 800-185](https://csrc.nist.gov/publications/detail/sp/800-185/final) defines KMAC[128,256](K, X, L, S) as a cSHAKE function. See the section 4.3 in NIST SP 800-185 for details. If KMAC is enabled, the software should configure [CMD.mode](https://docs.opentitan.org/hw/ip/kmac/doc/#Reg_cmd) to cSHAKE and the first six bytes of [PREFIX](https://docs.opentitan.org/hw/ip/kmac/doc/#Reg_prefix) to 0x01204B4D4143 (bigendian). The first six bytes of [PREFIX](https://docs.opentitan.org/hw/ip/kmac/doc/#Reg_prefix) represents the value of encode_string("KMAC")

Reply:
Maybe this should be a standalone DIF in itself. @timothytrippel?

Reply:
Sure I can add that. Will be a follow up PR though.

Reply:
we might want to hold off on the changes immediately, i think @moidx might be pointing out that this is undesireable behavior.  This potentially means that even during a keymgr derivation operation, another piece of software can influence the prefix used.

Reply:
Thanks @tjaychen for raising this. Since this is merged, I filed an issue #11751 to further discussion.